lib/phoenix/router.ex

```diff
@@ -293,8 +293,15 @@ defmodule Phoenix.Router do
       end
 
     quote do
-      defp do_call(unquote(conn), opts) do
-        unquote(call)
+      defp do_call(unquote(conn) = conn, opts) do
+        case conn.private[:phoenix_bypass] do
+          nil -> unquote(call)
+          {__MODULE__, pipeline} ->
```

Comment:
Here is what I would do: I would always store it as `{__MODULE__, pipelines}`. This way, we get to support both 0 and  many pipelines, all we need to do is to reduce over the connection:

    Plug.Conn.halt Enum.reduce pipelines, conn, &apply(__MODULE__, &1, [&2, []])

In the test helper, the default pipeline is an empty list, but we will also call `List.wrap` on the given argument so we can convert a single pipeline into a list.